Unicode support in format strings

sirrahd opened the pull request:
This change fixes an issue where the Datetime Format GUI fails to load with the error 'Failed to convert locale string to UTF8: Invalid byte sequence in conversion input' if unicode characters are entered in any of the format strings.

Date().toLocaleFormat(format) only accepts ascii characters in the format string. This change works around that limitation by escaping non-ascii characters in the format string before passing the string to toLocaleFormat(), then escaping the result in case non-ascii characters are in the localized string, and finally unescaping the resulting string by passing it to JSON.parse.

sirrahd committed:
Unicode support in format strings

This change fixes an issue where the Datetime Format GUI fails to load with the error 'Failed to convert locale string to UTF8: Invalid byte sequence in conversion input' if unicode characters are entered in any of the format strings.

Date().toLocaleFormat(format) only accepts ascii characters in the format string. This change works around that limitation by escaping non-ascii characters in the format string before passing the string to toLocaleFormat(), then escaping the result in case non-ascii characters are in the localized string, and finally unescaping the resulting string by passing it to JSON.parse.

Daniel-Khodabakhsh commented on src/Utilities.js:
```diff
@@ -12,7 +12,13 @@ const Gtk = imports.gi.Gtk;
 ///
 function dateTimeFormat(format, defaultFormat) {
-	return (format && new Date().toLocaleFormat(format) || format) || defaultFormat && new Date().toLocaleFormat(defaultFormat) || "";
+	function escapeUnicode(str) {
```
Can you move this function out and add comments like the other functions in this file? And could you also rename the 'str' parameter to 'string'?

Daniel-Khodabakhsh commented on src/Utilities.js:
```diff
@@ -12,7 +12,13 @@ const Gtk = imports.gi.Gtk;
+		});
+	}
+
+	return JSON.parse('"' + escapeUnicode((format && new Date().toLocaleFormat(escapeUnicode(format)) || format) || defaultFormat && new Date().toLocaleFormat(escapeUnicode(defaultFormat)) || "").replace('"',     '\\"') + '"');
```
Just wanted to test this before I merge it, want to make sure all edge cases are accounted for.

sirrahd commented:
Thanks Daniel. I did some edge case testing just now on this and found this doesn't handle " and \ characters well, need to escape them to make JSON.parse happy. I will push an update today with these fixes and your requested changes.

sirrahd committed:
Safe-escape " and \ characters, and update format.

This change expands on the previous unicode fix to also escape \ and " characters that are problematic for JSON.parse, and make formatting changes.

escapeJson() includes escaping problematic ascii characters and unicode, while escapeUnicode() escapes only unicode characters. escapeJson() should be run only once on the raw format string to avoid over-escaping, and the escapeUnicode() must be run after toLocaleFormat() because toLocaleFormat() can return additional unicode characters in languages like Japanese.

sirrahd commented:
Just pushed an update that addresses your feedback and fixes the " and \ issues.
